File: 002.rb
def fib(n)
 n < 2 ? n : fib(n-1) + fib(n-2) 
end

puts (1..Float::INFINITY).lazy.map { |i| fib(i) }.take_while { |i| i <= 4000000 }.select(&:even?).reduce(:+) # => 4613732

File: 002.scala
lazy val fibs: Stream[Int] = {
  def fib(a: Int, b: Int): Stream[Int] = a #:: fib(b, a + b)
  fib(0, 1)
}

println(fibs.takeWhile(_ <= 4000000).filter(_ % 2 == 0).sum) // => 4613732
